rspace/src/main/scala/coop/rchain/rspace/merger/instances/DiffStateMerger.scala

```diff
@@ -0,0 +1,317 @@
+
+      // main state data readers, used in computation for all mergings => so this is lazy cache
+      readProduces = (hash: Blake2b256Hash) => mainReader.getRichDatums(hash)
+      readConsumes = (hash: Blake2b256Hash) => mainReader.getRichContinuations(hash)
+      readJoin     = (hash: Blake2b256Hash) => mainReader.getRichJoins(hash)
+      mainProduces <- LazyKeyValueCache(readProduces)
+      mainConsumes <- LazyKeyValueCache(readConsumes)
+      mainJoins    <- LazyKeyValueCache(readJoin)
+
+      // map consume -> channels, used for optimisation
+      channelsPerConsumesRef <- Ref.of[F, Map[Blake2b256Hash, Seq[Blake2b256Hash]]](Map.empty)
+
+      // hash in history containing joins for channel
+      joinLeafForChannel <- LazyKeyValueCache(
+                             (hash: Blake2b256Hash) =>
+                               historyRepo
+                                 .getChannelHash(hash)
+                                 .flatMap {
+                                   case Some(DataJoinHash(_, j)) => j.pure[F]
+                                   case _ =>
+                                     Concurrent[F].raiseError[Blake2b256Hash](
+                                       new Exception(
+                                         s"hash $hash not found in channel store when requesting join"
+                                       )
+                                     )
+                                 }
+                           )
+
+      // each merging can be processed in parallel
+      work = fs2.Stream.emits(mergings.map(m => {
+        // history reader for end state
+        val startStateReader = historyRepo.getHistoryReader(m.startState.root)
+        val endStateReader   = historyRepo.getHistoryReader(m.endState.root)
+
+        // compute channels which are affected in merging
+        val (comms, prods, conss) =
+          m.events.foldLeft((Seq.empty[COMM], Seq.empty[Produce], Seq.empty[Consume]))(
+            (acc, v) =>
+              v match {
+                case c: COMM    => acc.copy(_1 = acc._1 :+ c)
+                case p: Produce => acc.copy(_2 = acc._2 :+ p)
+                case c: Consume => acc.copy(_3 = acc._3 :+ c)
+              }
+          )
+
+        // changes that history actions have to be computed for
+        val mergeChanges = {
+          // only produces and consumes that are not volatile in the scope of event log are subject to merge
+          // volatile events cannot have any effect outside if the scope
+          val newActiveProduces =
+            prods.filter(
+              p => p.persistent || !comms.exists(c => c.produces.contains(p) && c.peeks.isEmpty)
+            )
+          val newActiveConsumes = conss.filter(c => c.persistent || !comms.exists(_.consume == c))
+          val destroyedExternalProduces =
+            comms.filter(_.peeks.isEmpty).flatMap(_.produces).filterNot(p => prods.contains(p))
```

Comment:
what if the produce is persistent which should not be destroyed?

Reply:
Nice catch, we just do more work, but looks like it should be 
```          val destroyedExternalProduces =
            comms
              .filter(_.peeks.isEmpty)
              .flatMap(_.produces)
              .filterNot(p => prods.contains(p))
              .filterNot(_.persistent)
```